### ISSUE TITLE
Removing `findDOMNode`

### ISSUE THREAD

**Issue by rhys-vdw** [joshwcomeau/react-flip-move#89]
Hey @joshwcomeau. I look at #88 and saw [your comment](https://github.com/kdemoya/react-flip-move/blob/4ae798c45ca847daf7b33dc7465982f29aab7f35/src/dom-manipulation.js#L193-L195):

```
  // While ReactDOM's `findDOMNode` is discouraged, it's the only
  // publicly-exposed way to find the underlying DOM node for
  // composite components.
```

I've been reading up on `findDOMNode`, there's a discussion at yannickcr/eslint-plugin-react#678. I thought I'd just leave a comment here detailing how one might remove use of this function. I'm not necessarily convinced this is a good idea.

First option is `FlipMove` would expect each child to be a html element, adding a `ref` with `cloneElement`. This means library consumers will have to wrap custom components in an `li` or `div`. `FlipMove` would just assume that all instances returned by `ref` are elements (erroring/warning otherwise).

Another option might be to require the custom child components to expose a `containerRef` prop.

```js
// `containerRef` prop required/supplied by `FlipMove`
function Custom({ containerRef, text }) {
  return <li ref={containerRef}>{text}</li>
}

function List({ items }) {
  const itemElements = items.map(item =>
    <Custom text={item.text} />
  )
  return <FlipMove>{itemElements}</FlipMove>
}
```

FlipMove can do something like:

```js
  childrenWithRefs() {
    return this.state.children.map(child => {
      const ref = domNode => {
        if (!this.childrenData[child.key]) {
          this.childrenData[child.key] = {};
        }
        this.childrenData[child.key].domNode = domNode;
      }
      // If it's a html element, then we just use `ref`, otherwise require `containerRef`.
      return isHtmlElement(child.type)
        ? React.cloneElement(child, { ref })
        : React.cloneElement(child, { containerRef: ref })
    })
  }
```

Nice thing about this approach is that it means you can use stateless function components since they just have to pass the `ref` up via props.

**Comment by joshwcomeau** [comment]
Hey man,

Yeah, I had an internal debate about this myself :) I really don't wanna be reliant on a soon-to-be deprecated API, but I also really don't wanna complicate integration instructions. 

Having users specify a `callbackRef` is a pretty big ask; I don't wanna impose on people new to React who aren't familiar with refs and how they work. It can also be a fair bit of trouble if you have several layers of composite components.

I do really like that it would enable SFC's, though.

I think for now, I'd like to keep it as-is. If/when `findDOMNode` becomes officially deprecated, I think that'll be the time to make the switch.

We can also enable support for SFCs in the meantime by allowing users to "opt-in" voluntarily, and using it if it's supplied! Would gladly accept a PR that enables this. I think in terms of its name, it should be something like `animationRef` or `flipRef`? Just to make it clear _why_ it's needed.

BTW, not sure if you saw, but I [refactored the codebase](https://github.com/joshwcomeau/react-flip-move/pull/83/files) pretty heavily. My coding style has changed quite a bit in the last year, and Flip Move was due for some housekeeping. I think it'll make it much nicer to work on :) I still need to redo the tests, they're still pretty messy.

**Comment by rhys-vdw** [comment]
> I think for now, I'd like to keep it as-is. If/when findDOMNode becomes officially deprecated, I think that'll be the time to make the switch.

Agreed. The purpose of this issue was really just to jot down what I learned.

> We can also enable support for SFCs in the meantime by allowing users to "opt-in" voluntarily

Unless somebody requests it probably not worth the added code and documentation complexity IMO. It's simpler (and more readable) to just wrap them in `li`/`div` components. We can close this issue if you like, or leave it as a placeholder for discussion if `findDOMNode` is deprecated.

**Comment by geelen** [comment]
Personally, I think enforcing that this only works with a layer of real HTML elements makes a lot of sense. It feels like it almost demystifies the library because it's clearly just working at this one layer of the DOM tree, rather than magically working with class components and not working at all with SFCs.

I'm going to start doing that by convention anyway 👍

**Comment by Hypnosphi** [comment]
What do you guys think about wrapping every child in a div (or an optional user-provided tag) inside FlipMove's render, as Dan Abramov [suggests](https://github.com/yannickcr/eslint-plugin-react/issues/678#issuecomment-232293175) for a similar case of HOCs? This off course would be a breaking change, but looks like it may become the only suitable option eventually, when [fragment API](https://github.com/facebook/react/issues/2127) ships

**Comment by joshwcomeau** [comment]
> What do you guys think about wrapping every child in a div (or an optional user-provided tag) inside FlipMove's render, as Dan Abramov suggests for a similar case of HOCs? This off course would be a breaking change, but looks like it may become the only suitable option eventually, when fragment API ships

I was actually thinking about this myself a few days ago. The big caveat is that it would cause even more problems for people when the markup is important (grids, flexbox, tables, etc).

For fragments, we could insist that, for direct children of FlipMove, a single top-level child be returned... Since, even if we do wrap it in a `<span>`, we'd be wrapping _all_ of the children, instead of wrapping each child individually.

I'm unsure of the fragment API implementation; is it the case that _all_ component renders are treated as fragments, or only those that return an array?

**Comment by Hypnosphi** [comment]
>we'd be wrapping all of the children

I think it's actually an expected behaviour, as long as we don't embrace #148's approch and keep using react children as animated entities. The component's representation being more than one dom element is just an implementation detail

>is it the case that all component renders are treated as fragments, or only those that return an array?

Only the latter

**Comment by Hypnosphi** [comment]
>The big caveat is that it would cause even more problems for people when the markup is important (grids, flexbox, tables, etc)

For such cases, if just redefining the wrapper's tagname isn't sufficient, there should be a possibility to opt out of implicit wrapping. In that case, user should explicitly wrap each child into some dom element as suggested above
